lib/turbotrig/turbovec.c

```diff
@@ -0,0 +1,238 @@
+#pragma GCC diagnostic ignored "-Wstrict-aliasing"
+
+#include <breezystm32/breezystm32.h>
+
+#include "turbovec.h"
+
+static void pfvec(vector_t v)
+{
+  printf("[%d, %d, %d]\n", (int32_t)(v.x*1000), (int32_t)(v.y*1000), (int32_t)(v.z*1000));
+}
+
+static void pfquat(quaternion_t v)
+{
+  printf("[%d, %d, %d, %d]\n", (int32_t)(v.w*1000), (int32_t)(v.x*1000), (int32_t)(v.y*1000), (int32_t)(v.z*1000));
+}
+
+
+
+int32_t int_dot(intvec_t v1, intvec_t v2)
+{
+  return (v1.x*v2.x + v1.y*v2.y + v1.z*v2.z)/1000;
+}
+
+intvec_t int_cross(intvec_t u, intvec_t v)
+{
+  intvec_t out;
+  out.x = (u.y*v.z - u.z*v.y)/1000;
+  out.y = (u.z*v.x - u.x*v.z)/1000;
+  out.z = (u.x*v.y - u.y*v.x)/1000;
+  return out;
+}
+
+intvec_t int_scalar_multiply(int32_t s, intvec_t v)
+{
+  intvec_t out;
+  out.x = (s*v.x)/1000;
+  out.y = (s*v.y)/1000;
+  out.z = (s*v.z)/1000;
+  return out;
+}
+
+intvec_t int_vector_add(intvec_t u, intvec_t v)
+{
+  intvec_t out;
+  out.x = u.x + v.x;
+  out.y = u.y + v.y;
+  out.z = u.z + v.z;
+  return out;
+}
+
+intvec_t int_vector_sub(intvec_t u, intvec_t v)
+{
+  intvec_t out;
+  out.x = u.x - v.x;
+  out.y = u.y - v.y;
+  out.z = u.z - v.z;
+  return out;
+}
+
+int32_t int_sqrd_norm(intvec_t v)
+{
+  return v.x*v.x + v.y*v.y + v.z*v.z;
+}
+
+intvec_t int_vector_normalize(intvec_t v)
+{
+  float recipNorm = turboInvSqrt((v.x*v.x + v.y*v.y + v.z*v.z)/1000000.0);
+  return int_scalar_multiply((int32_t)(1000*recipNorm), v);
+}
+
+intquat_t int_quaternion_normalize(intquat_t q)
+{
+  float recipNorm = turboInvSqrt((q.w*q.w + q.x*q.x + q.y*q.y + q.z*q.z) / 1000000.0);
+  int32_t intRecipNorm = (int32_t)(recipNorm*1000);
+  intquat_t out;
+  out.w = (intRecipNorm*q.w)/1000;
+  out.x = (intRecipNorm*q.x)/1000;
+  out.y = (intRecipNorm*q.y)/1000;
+  out.z = (intRecipNorm*q.z)/1000;
+  return out;
+}
+
+intquat_t int_quaternion_multiply(intquat_t q1, intquat_t q2)
+{
+  int32_t s1 = q1.w;
+  int32_t s2 = q2.w;
+
+  intvec_t v1 = {q1.x, q1.y, q2.z};
+  intvec_t v2 = {q2.x, q2.y, q2.z};
+
+  int32_t w = (s1*s2)/1000 - int_dot(v1, v2);
+  // xyz = s1*v2 + s2*v1 - v1 x v2)
+  intvec_t xyz = int_vector_sub(int_vector_add(int_scalar_multiply(s1, v2), int_scalar_multiply(s2,v1)), int_cross(v1,v2));
+  intquat_t q = {w, xyz.x, xyz.y, xyz.z};
+  return int_quaternion_normalize(q);
+}
+
+
+intquat_t int_quaternion_inverse(intquat_t q)
+{
+  intquat_t out = {-1*q.x, q.y, q.x, q.z};
```

Comment:
Either works.  One is reversing the angle, the other reverses the vector. I haven't actually tried both ways, but I expect that it would result in the same thing